### ISSUE TITLE
[10k]: Serialization and API documentation planning

### ISSUE THREAD

**Issue by widal001** [HHS/simpler-grants-gov#66]
### Description

Make an analysis of needs and alternatives and choose:

- Serialization framework (e.g., Marshmallow or alternative)
- API autogenerator
- API documentation (e.g., OpenAPI) autogenerator
- Database migration tool (e.g., Alembic)

Ideally, fields would be defined only once in the codebase, and these definitions would be used to generate (with as little manual configuration as possible) both RESTful API endpoints and the documentation for those endpoints.

Otherwise, every time there is a change in schema, it requires a lot of developer attention to update field definitions, API definitions, and API documentation.

### Dependencies

- [x] #42 

### Sub-Tasks

Sub-tasks can be found in [this milestone](https://github.com/HHS/simpler-grants-gov/milestone/98)

- [x] [ADR] Serialization framework - https://github.com/HHS/simpler-grants-gov/issues/765
- [x] [ADR] API specification framework
- [x] [ADR] Database migration tool - https://github.com/HHS/simpler-grants-gov/issues/765

### Definition of Done

- [x] An ADR has been created for each of the tools chosen above:
  - [x] Serialization framework - #28 
  - [x] API specification
  - [x] Database migrations - https://github.com/HHS/simpler-grants-gov/issues/765
- [x] We've documented a proposed strategy for keeping changes between data model, response schemas, and API documents in sync 
   - [x] Documentation on how it works

**Comment by acouch** [comment]
We discussed with @widal001 and @chouinar . We decided that there isn't a need to create an ADR for serialization or API doc generation, as that is mostly covered in the [Flask + FlaskAPI ADR](). We did identity a couple of tasks:

* Document common tasks or patterns in API documtentation
    * add new column to the codebase
    * add fields to the API
* Update ADR to give a little more detail on pros / cons serialization and schema changes
* Noting that we want the dev-centered documentation to be mentally consistent across apps

**Comment by widal001** [comment]
### Update 

With the following documentation that @chouinar and other members of the engineering team have created, this ticket is "done" from my perspective. If @lucasmbrown-usds feels like something is missing we can either re-open it or create a separate follow-up ticket.

### Documentation

- [API README](https://github.com/HHS/simpler-grants-gov/tree/main/api#introduction)
- [API repo docs](https://github.com/HHS/simpler-grants-gov/tree/main/documentation/api)
- API-related ADRs
  - [API framework](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/decisions/adr/2023-07-07-api-framework.md)
  - [Database](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/decisions/adr/2023-07-05-db-choices.md)
  - [Backend tooling](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/decisions/adr/2023-07-07-backend-tooling.md)
  - [Backend production server](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/decisions/adr/2023-07-26-backend-prod-server.md)
  - [Database migrations](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/decisions/adr/2023-12-06-database-migrations.md)
  - [Uptime monitoring](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/decisions/adr/2023-11-22-uptime-monitoring.md)

### Notes

On the following acceptance criterion:
> We've documented a proposed strategy for keeping changes between data model, response schemas, and API documents in sync

We wanted to note that there are two main areas of synchronization we want to maintain:
- **DB model in the API <> DB schema in production** - Our use of alembic meets this need and the docs that Michael put together in database sub-directory do a great job of explaining how to work with DB migrations.
- **Response model in the API <> OpenAPI docs** - Our use of marshmallow + APIFlask satisfies this need, and the api-details docs do a good job of highlighting the relationship between serialization schemas and OpenAPI docs.

**Comment by sumiat** [comment]
More details in this Slack thread: https://betagrantsgov.slack.com/archives/C05TSL64VUH/p1703270115707139
